up_client_socket/rust/utransport-socket/src/lib.rs

```diff
@@ -0,0 +1,420 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+ mod constants;
+
+ use async_trait::async_trait;
+
+ use up_rust::UListener;
+ use up_rust::{UAttributesValidators, UriValidator};
+ use up_rust::{UCode, UMessage, UMessageType, UStatus, UTransport, UUri};
+ 
+ use crate::constants::BYTES_MSG_LENGTH;
+ use crate::constants::DISPATCHER_ADDR;
+ use protobuf::Message;
+ use std::collections::hash_map::Entry;
+ use std::collections::HashSet;
+ use std::io::{Read, Write};
+ use std::net::TcpStream as TcpStreamSync;
+ use std::{
+     collections::HashMap,
+     sync::{Arc, Mutex},
+ };
+ use tokio::task;
+ use up_rust::ComparableListener;
+ 
+ pub struct UTransportSocket {
+     socket_sync: TcpStreamSync,
+     listener_map: Arc<Mutex<HashMap<UUri, HashSet<ComparableListener>>>>,
+ }
+ impl Clone for UTransportSocket {
+     fn clone(&self) -> Self {
+         UTransportSocket {
+             socket_sync: self
+                 .socket_sync
+                 .try_clone()
```

Comment:
IMHO this is again a bit of an odd construction. Suggest revisiting the refactoring I discussed last time.